### PR TITLE
Refactor SGD __decayed_lr to allow tf.Variable to be used as decay parameter

### PR DESCRIPTION
**Pull request by andyclegg**
This PR refactors the `__decayed_lr` implementation in SGD to use `tf.cond` and `tf.greater` rather than the native Python versions, in order to allow a `tf.Variable` to be used as the decay parameter.

### CODE REVIEW

**Comment by andyclegg** [comment]
This is for issue #15934

**Comment by fchollet** [comment]
Thanks for the PR. What is the use case for this? I would actually recommend using a [learning rate schedule](https://keras.io/api/optimizers/learning_rate_schedules/) object to control the learning rate decay. Using a mutable variable would create a rather confusing UX.

**Comment by andyclegg** [comment]
The use case is for hyperparameter searching. We would ideally like to keep the graph the same between runs with different values for decay (including `decay=0`).

**Comment by fchollet** [comment]
> The use case is for hyperparameter searching. We would ideally like to keep the graph the same between runs with different values for decay (including decay=0).

Doing hyperparameter search via mutable variables in a static graph makes for a confusing mental model. Any component of your model may be doing something like `a = int(a)` in a constructor (for example), which would lead that component to use the value of `a` at instantiation time and ignore any subsequent update to the value of the variable `a`. I would strongly discourage you from going this route as you will likely face hard-to-detect bugs in the future.

It seems safer to disallow TF variables as hyperparameters altogether. We should keep a clear distinction between "static graph hyperparameters" (which are fixed for a given graph) and "mutable state". Mutable state typically only includes layer model weights, optimizer weights, and the learning rate of the optimizer. Decay so far has been treated as static.

**Comment by andyclegg** [comment]
@fchollet - that's fair enough. Thanks for taking the time to explain! I'll close the PR and associated issue.
